File: [0428][Hard][Serialize_and_Deserialize_N-ary_Tree]/Serialize_and_Deserialize_N-ary_Tree.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 80 ms
# leetcode memory   cost : 16.8 MB
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
# solution 1, DFS
class Codec:

    def serialize(self, root: 'Node') -> str:
        """Encodes a tree to a single string.
        
        :type root: Node
        :rtype: str
        """
        res = []  

        def dfs(root):
            if not root: return 
            res.append(str(root.val))
            res.append(str(len(root.children)))
            for ch in root.children:
                dfs(ch)
        
        dfs(root)
        # print(res)
        return ",".join(res)

    def deserialize(self, data: str) -> 'Node':
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: Node
        """
        if not data: return 
        data = iter(data.split(","))
        
        def dfs():
            root = Node(int(next(data)), [])
            num = int(next(data))
            for _ in range(num):
                root.children.append(dfs())
            return root
        
        return dfs()
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: [0428][Hard][Serialize_and_Deserialize_N-ary_Tree]/Serialize_and_Deserialize_N-ary_Tree.cpp
// leetcode time     cost : 108 ms
// leetcode memory   cost : 199.2 MB 
#include <vector>
#include <stdlib.h>
#include <string>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(Node* root) {
        string s;
        if (!root) {
            s = "#";
            return s;
        }
        s = to_string(root->val) + " " + to_string(root->children.size()); 
        for (const auto &child : root->children) {
            s = s + " " + serialize(child);
        }
        return s;
    }

    // Decodes your encoded data to tree.
    Node* deserialize(string data) {
        istringstream is(data);
        return DFS(is);
    }

    Node* DFS(istringstream &is) {
        string s, size;
        is >> s;
        if (s == "#") return nullptr;
        is >> size;
        Node *p = new Node(stoi(s));
        for (int i = 0; i < stoi(size); ++i) {
            p->children.push_back(DFS(is));
        }
        return p;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));